pkg/agent/monitortool/latency_store.go

```diff
@@ -0,0 +1,250 @@
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+
+	delete(l.nodeIPLatencyMap, nodeIP)
+}
+
+// SetNodeIPLatencyEntry sets the NodeIPLatencyEntry for the given Node IP
+func (l *LatencyStore) SetNodeIPLatencyEntry(nodeIP string, mutator func(entry *NodeIPLatencyEntry)) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+
+	entry, ok := l.nodeIPLatencyMap[nodeIP]
+	if !ok {
+		entry = &NodeIPLatencyEntry{}
+		l.nodeIPLatencyMap[nodeIP] = entry
+	}
+
+	mutator(entry)
+}
+
+// addNode adds a Node to the latency store
+func (l *LatencyStore) addNode(node *corev1.Node) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+
+	l.updateNodeMap(node)
+}
+
+// deleteNode deletes a Node from the latency store
+func (l *LatencyStore) deleteNode(node *corev1.Node) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+
+	nodeIPs, ok := l.nodeTargetIPsMap[node.Name]
+	if !ok {
+		klog.ErrorS(nil, "Failed to get IPs for Node", "nodeName", node.Name)
+		return
+	}
+
+	for _, ip := range nodeIPs {
+		delete(l.nodeIPLatencyMap, ip.String())
+	}
+
+	delete(l.nodeTargetIPsMap, node.Name)
+}
+
+// updateNode updates a Node name in the latency store
+func (l *LatencyStore) updateNode(new *corev1.Node) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+
+	// Node name will not be changed in the same Node update operation.
+	l.updateNodeMap(new)
+}
+
+// updateNodeMap updates the nodeTargetIPsMap with the IPs of the given Node.
+func (l *LatencyStore) updateNodeMap(node *corev1.Node) {
+	nodeIPs, err := l.getNodeIPs(node)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get IPs for Node", "nodeName", node.Name)
+		return
+	}
+
+	l.nodeTargetIPsMap[node.Name] = nodeIPs
+}
+
+// getNodeIPs returns the target IPs of the given Node based on the agent mode.
+func (l *LatencyStore) getNodeIPs(node *corev1.Node) ([]net.IP, error) {
+	if l.isNetworkPolicyOnly {
+		transportIPs, err := getTransportIPs(node)
+		if err != nil {
+			return nil, err
+		}
+
+		return transportIPs, nil
+	} else {
+		gw0IPs, err := getGWIPs(node)
+		if err != nil {
+			return nil, err
+		}
+
+		return gw0IPs, nil
+	}
+}
+
+// getTransportIPs returns the transport IPs of the given Node.
+func getTransportIPs(node *corev1.Node) ([]net.IP, error) {
+	var transportIPs []net.IP
+	ips, err := k8s.GetNodeTransportAddrs(node)
+	if err != nil {
+		return transportIPs, err
+	}
+
+	if ips.IPv4 != nil {
+		transportIPs = append(transportIPs, ips.IPv4)
+	}
+	if ips.IPv6 != nil {
+		transportIPs = append(transportIPs, ips.IPv6)
+	}
+
+	return transportIPs, nil
+}
+
+// getGWIPs returns the gateway IPs of the given Node.
+func getGWIPs(node *corev1.Node) ([]net.IP, error) {
+	var gwIPs []net.IP
+
+	podCIDRStrs := getPodCIDRsOnNode(node)
+	if len(podCIDRStrs) == 0 {
+		// Skip the Node if it does not have a PodCIDR.
+		err := errors.New("node does not have a PodCIDR")
+		return nil, err
+	}
+
+	// the 0th entry must match the podCIDR field. It may contain at most 1 value for
+	// each of IPv4 and IPv6.
```

Comment:
remove: this comment doesn't make sense here because we are not using the `podCIDR` field (it was accessed in `getPodCIDRsOnNode`).

Reply:
Okay, I've fixed that.